Lib/test/test_warnings/__init__.py

```diff
@@ -472,6 +473,42 @@ def test_stacklevel_import(self):
                 self.assertEqual(len(w), 1)
                 self.assertEqual(w[0].filename, __file__)
 
+    def test_skip_file_prefixes(self):
+        with warnings_state(self.module):
+            with original_warnings.catch_warnings(record=True,
+                    module=self.module) as w:
+                self.module.simplefilter('always')
+
+                # Warning never attributed to the data/ package.
+                package_helper.inner_api(
+                        "inner_api", stacklevel=2,
+                        warnings_module=warning_tests.warnings)
+                self.assertEqual(w[-1].filename, __file__)
+                warning_tests.package("package api", stacklevel=2)
+                self.assertEqual(w[-1].filename, __file__)
+                self.assertEqual(w[-2].filename, w[-1].filename)
+                # Low stacklevels are overridden to 2 behavior.
+                warning_tests.package("package api 1", stacklevel=1)
+                self.assertEqual(w[-1].filename, __file__)
+                warning_tests.package("package api 0", stacklevel=0)
+                self.assertEqual(w[-1].filename, __file__)
+                warning_tests.package("package api -99", stacklevel=-99)
+                self.assertEqual(w[-1].filename, __file__)
+
+                # The stacklevel still goes up out of the package.
+                warning_tests.package("prefix02", stacklevel=3)
+                self.assertIn("unittest", w[-1].filename)
+
+    def test_skip_file_prefixes_type_errors(self):
+        with warnings_state(self.module):
+            warn = warning_tests.warnings.warn
+            with self.assertRaises(TypeError):
+                warn("msg", skip_file_prefixes=[])
+            with self.assertRaises(TypeError):
+                warn("msg", skip_file_prefixes=(b"bytes",))
```

Comment:
I take it file paths from frames can't be bytes paths?

Reply:
While I didn't see it explicitly mentioned anywhere, signs point to "`code->co_filename` is always a unicode `str`" based on what I see being done with in some places in the CPython code. For example `Python/traceback.c` `tb_printinternal` passes it to `tb_displayline` which references it using `%U` in `PyUnicode_FromFormat`... that would explode if it were not.  `Modules/_tracemalloc.c` also balks if co_filename fails `PyUnicode_Check`, but things like code_repr and the existing code in `_warnings.c` punt if `PyUnicode_Check` fails. So it's more that this feature is unicode filenames only, not implementing `bytes` or arbitrary objects supporting a `.prefix` API match. This appears to be good enough for all know uses.